### DISCUSSION TITLE
Philosophy: Machines with dynamic configuration

### DISCUSSION THREAD

**Discussion by JayNaire** [statelyai/xstate#2062] [General]
I'd appreciate any pointers/papers on dealing with systems/machines whose structure changes with time.

I'm building a central heating system.
Real timish control is by transforming events from sensors to outputs to actuators.
Both sensors and actuators can, infrequently, come and go during normal operation (nature of the beast).
On startup a machine is made from arrays of currently known devices*. A supervisor maintains these arrays and generates events when they change.)
My intention, when something changes, is to stop the current machine, build a new one and continue.
Something doesn't smell right about this approach but it would stop having to write a lot of reversionary code.

Fresh insights gratefully received!

*example:
```
...
const valves = ["a", "b", "c", "d" ]; /*maintained globally*/
const controlStates = ["off", "ready", "timeout"]
function makeStates() {
  let allStates = controlStates.reduce((allControlStates, ctl) => {
    allControlStates[ctl] = {
      meta: { name: ctl },
      on: valves.reduce((acc, valve) => {
        acc[(valve)] = {
          target: valve
        };
        return acc
      }, {}),
      activities: ["ticker"]
    }
    return allControlStates
  }, {})

  valves.reduce((allValves, key) => {
    allStates[key] = {
      meta: { name: key },
      entry: ["sendRequests"],
      on: valves.reduce((acc, valve) => {
        acc[(valve + "_RESPONSE")] = {
          actions: ["responseArrived"]
        };
        acc["_READY"] = "ready";
        acc["_TIMEOUT"] = "timeout";
        if (valve !== key) {
          acc[valve] = {
            target: valve
          };
        }
        return acc;
      }, {}),
      activities: ["ticker", "awaitReplies"]
    };
    return allValves;
  }, {})

  return allStates
}//makeStates
```
...

**Comment by davidkpiano** [comment]
> My intention, when something changes, is to stop the current machine, build a new one and continue.
> Something doesn't smell right about this approach but it would stop having to write a lot of reversionary code.

Yeah, this doesn't feel right to me either. My recommendation would be to reify that "when something changes" part into an actual concrete event. Then determine how your machine should behave due to changes, without stopping/restarting the machine.

Also, try to use `invoke` instead of `activities`, which will be deprecated in v5.

**Reply by JayNaire** [reply]
Thanks @davidkpiano. I only wish you had disagreed with me!
I appreciate the invoke tip.

Sometimes I feel that designing state machines is like trying to play the Chopin Fantaisie Impromptu when I've barely mastered Chopstix.
